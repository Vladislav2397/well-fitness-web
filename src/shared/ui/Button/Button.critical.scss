.button {
    font-size: toRem(14);
    line-height: toRem(16);
    font-weight: 600;
    border-radius: toRem(4);
    border: 1px solid transparent;
    vertical-align: middle;

    i {
        vertical-align: middle;
    }

    &--size {
        &-s {
            padding: toRem(8);
        }

        &-m {
            padding: toRem(11) toRem(19) toRem(12);
        }
    }

    &--theme {
        &-brand {
            background-color: #f53b49;
            color: #ffffff;
            border-color: #f53b49;
        }

        &-secondary {
            background-color: #909cb5;
            border-color: #909cb5;
            color: #ffffff;
        }

        &-ghost-brand {
            color: #f53b49;
            border-color: #f53b49;
        }

        &-ghost {
            border-color: #909cb5;
            color: #909cb5;
        }

        &-ghost-light {
            border-color: $white;
            color: $white;
        }
    }
}
