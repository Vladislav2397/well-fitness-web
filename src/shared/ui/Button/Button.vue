<script lang="ts">
import { CreateElement, VNode } from 'vue'
import { Component, Emit, Prop, PropSync, Vue } from 'vue-property-decorator'

export type ButtonProps = {
    tag: 'button' | 'a' | 'div' | 'router-link'
    iconSize: 's' | 'm' | 'l'
    size: 's' | 'm'
    theme: buttonThemeType
    icon: string
    href?: string
    errorSync?: boolean
    error?: boolean
    iconLeft: boolean
}

export type buttonThemeType =
    | 'brand'
    | 'ghost-brand'
    | 'ghost-light'
    | 'ghost'
    | 'secondary'

@Component
export default class Button extends Vue implements ButtonProps {
    @Prop({ default: 'button' }) readonly tag!:
        | 'button'
        | 'a'
        | 'div'
        | 'router-link'
    @Prop({ default: 'm' }) readonly size!: 'm' | 's'
    @Prop({ default: 'brand' }) readonly theme!: ButtonProps['theme']
    @Prop() readonly icon!: string
    @Prop({ default: 'm' }) readonly iconSize!: ButtonProps['iconSize']
    @Prop() readonly href!: string
    @Prop() readonly iconLeft!: boolean

    @PropSync('error') errorSync!: boolean

    @Emit('click') clickEmit(): void {
        /**/
    }

    get classes(): string[] {
        const classes = ['button']

        classes.push(`button--theme-${this.theme}`)
        classes.push(`button--size-${this.size}`)

        if (this.icon) {
            if (this.iconLeft) {
                classes.push(`button--icon-left`)
            } else {
                classes.push(`button--icon-right`)
            }
        }

        return classes
    }

    render(h: CreateElement): VNode {
        const children = []

        const icon = h('i', {
            class: `icon icon--size-${this.iconSize} icon--${this.icon}`,
        })

        if (this.icon) {
            if (this.iconLeft) {
                children.push(icon)
            }
            children.push(this.$slots.default)
            if (!this.iconLeft) {
                children.push(icon)
            }
        } else {
            children.push(this.$slots.default)
        }

        return h(
            'button',
            {
                class: this.classes,
                ...(this.tag === 'a' && {
                    attrs: {
                        href: this.href,
                        target: '_blank',
                        rel: 'noopener noreferrer',
                    },
                }),
                ...(this.tag === 'router-link' && {
                    attrs: {
                        to: this.href,
                    },
                }),
                on: {
                    click: this.clickEmit,
                },
            },
            children,
        )
    }
}
</script>

<!--<style lang="scss" src="./Button.critical.scss" />-->
