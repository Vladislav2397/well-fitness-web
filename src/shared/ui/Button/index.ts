import Button, { type ButtonThemeType } from './Button.vue'

export { Button, ButtonThemeType as buttonThemeType }
