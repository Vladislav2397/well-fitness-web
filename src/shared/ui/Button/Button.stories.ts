import { Story } from '@storybook/vue'
import { action } from '@storybook/addon-actions'

import Button from './Button.vue'
import { icons } from '@/shared/ui/Icon'
import { defineControls } from '@/shared/lib/stories'
import { ButtonProps } from '@/shared/ui/Button/Button.vue'

export default {
    title: 'shared/Button',
    component: Button,
    argTypes: defineControls({
        tag: ['button', 'a', 'div', 'router-link'],
        size: ['m', 's'],
        theme: ['brand', 'ghost-brand', 'ghost', 'secondary'],
        icon: ['', ...icons],
        iconSize: ['s', 'm', 'l'],
        href: '',
        iconLeft: false,
    }),
}

const Template: Story<ButtonProps> = (args, { argTypes }) => ({
    props: Object.keys(argTypes),
    components: { 'button-component': Button },
    methods: {
        handle: action('clicked'),
    },
    template: `<button-component v-bind="$props" @click="handle">Button</button-component>`,
})

export const Default = Template.bind({})
Default.args = {
    //
}
