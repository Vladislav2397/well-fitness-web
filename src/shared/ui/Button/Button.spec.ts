import { shallowMount } from '@vue/test-utils'
import Button from './Button.vue'

const href = 'https://google.com'
const to = '/'

describe('Button', () => {
    it('default snapshot', () => {
        const wrapper = shallowMount(Button)
        expect(wrapper.html()).toMatchSnapshot()
    })
    it.each([
        { tag: 'button' },
        { tag: 'div' },
        { tag: 'a', href },
        { tag: 'router-link', to },
    ])('with tag', ({ tag, href, to }) => {
        const wrapper = shallowMount(Button, {
            propsData: {
                tag,
                href,
                to,
            },
            stubs: ['router-link'],
            slots: {
                default: 'Button',
            },
        })
        expect(wrapper.html()).toMatchSnapshot()
    })
    it('as link', () => {
        const wrapper = shallowMount(Button, {
            propsData: {
                tag: 'a',
                href,
            },
        })
        expect(wrapper.html()).toMatchSnapshot()
    })
})
