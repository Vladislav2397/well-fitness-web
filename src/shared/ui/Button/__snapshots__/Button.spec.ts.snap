// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button all tags 1`] = `<button class="button button--theme-brand button--size-m">Button</button>`;

exports[`Button all tags 2`] = `<div class="button button--theme-brand button--size-m">Button</div>`;

exports[`Button all tags 3`] = `<a href="https://google.com" target="_blank" rel="noopener noreferrer" class="button button--theme-brand button--size-m">Button</a>`;

exports[`Button all tags 4`] = `<router-link-stub class="button button--theme-brand button--size-m" to="/">Button</router-link-stub>`;

exports[`Button as link 1`] = `<a href="https://google.com" target="_blank" rel="noopener noreferrer" class="button button--theme-brand button--size-m"></a>`;

exports[`Button default snapshot 1`] = `<button class="button button--theme-brand button--size-m"></button>`;

exports[`Button with tag 1`] = `<button class="button button--theme-brand button--size-m">Button</button>`;

exports[`Button with tag 2`] = `<div class="button button--theme-brand button--size-m">Button</div>`;

exports[`Button with tag 3`] = `<a href="https://google.com" target="_blank" rel="noopener noreferrer" class="button button--theme-brand button--size-m">Button</a>`;

exports[`Button with tag 4`] = `<router-link-stub class="button button--theme-brand button--size-m" to="/">Button</router-link-stub>`;
